cms/private_media/signal_handlers.py

```diff
@@ -0,0 +1,140 @@
+    return None
+
+
+def publish_media_on_page_publish(instance: "Page", **kwargs: Any) -> None:
+    """Signal handler to be connected to the 'page_published' signal for
+    all page types. It is responsible for identifying any privacy-controlled
+    media used by the page, and ensuring that it is also made public.
+    """
+    if not isinstance(instance, MediaParentMixin):
+        return None
+
+    for model_class in get_parent_derived_privacy_models():
+        queryset = model_class.objects.filter(
+            is_private=True, parent_object_content_type=instance.cached_content_type, parent_object_id=instance.id
+        )
+        if issubclass(model_class, AbstractImage):
+            queryset = queryset.prefetch_related("renditions")
+        if hasattr(model_class.objects, "bulk_make_public"):
+            model_class.objects.bulk_make_public(queryset)
+        else:
+            raise ImproperlyConfigured(
+                f"The default manager for {model_class.__name__} does not have a 'bulk_make_private' method."
+            )
+
+    return None
+
+
+def unpublish_media_on_page_unpublish(instance: "Page", **kwargs: Any) -> None:
+    """Signal handler to be connected to the 'page_unpublished' signal for
+    all page types. It is responsible for identifying any privacy-controlled
+    media used exclusively by the page, and ensuring that it is also made
+    private.
+    """
+    if not isinstance(instance, MediaParentMixin):
+        return None
+
+    for model_class in get_parent_derived_privacy_models():
+        queryset = model_class.objects.filter(
+            is_private=False, parent_object_content_type=instance.cached_content_type, parent_object_id=instance.id
+        )
+        if issubclass(model_class, AbstractImage):
+            queryset = queryset.prefetch_related("renditions")
+        queryset = queryset.annotate(
+            is_referenced_elsewhere=Exists(
+                ReferenceIndex.objects.filter(
+                    to_content_type=ContentType.objects.get_for_model(model_class),
+                    to_object_id=Cast(OuterRef("pk"), output_field=CharField()),
+                ).exclude(content_type=instance.cached_content_type, object_id=str(instance.id))
+            )
+        ).filter(is_referenced_elsewhere=False)
+        if hasattr(model_class.objects, "bulk_make_private"):
+            model_class.objects.bulk_make_private(queryset)
+        else:
+            raise ImproperlyConfigured(
+                f"The default manager for {model_class.__name__} does not have a 'bulk_make_private' method."
+            )
+    return None
+
+
+def register_signal_handlers() -> None:
+    """Register signal handlers for models using the private media system."""
+    page_published.connect(publish_media_on_page_publish, dispatch_uid="publish_media")
+    page_unpublished.connect(unpublish_media_on_page_unpublish, dispatch_uid="unpublish_media")
```

Comment:
NOTE: Wagtail also has non-page versions of these signals (`published` and `unpublished`, respectively) that we can use for parent 'Snippet' models. But, there aren't any existing snippets with media yet to try those out on.